src/components/lab/NumberInput/NumberInput.tsx

```diff
@@ -0,0 +1,326 @@
+        const delta = e.shiftKey ? e.deltaX : e.deltaY;
+        e.preventDefault();
+        if (delta > 0) {
+            handleIncrement(e);
+        } else if (delta < 0) {
+            handleDecrement(e);
+        }
+    };
+
+    const handleKeyDown: React.KeyboardEventHandler<HTMLInputElement> = (e) => {
+        if (e.key === KeyCode.ARROW_DOWN) {
+            e.preventDefault();
+            handleDecrement(e);
+        } else if (e.key === KeyCode.ARROW_UP) {
+            e.preventDefault();
+            handleIncrement(e);
+        } else if (e.key === KeyCode.HOME) {
+            e.preventDefault();
+            if (min !== undefined) {
+                setValue?.(min);
+                setInputValue(min.toString());
+            }
+        } else if (e.key === KeyCode.END) {
+            e.preventDefault();
+            if (max !== undefined) {
+                const newValue = clampToNearestStepValue({
+                    value: max,
+                    step: baseStep,
+                    min,
+                    max,
+                });
+                setValue?.(newValue);
+                setInputValue(newValue.toString());
+            }
+        }
+        onKeyDown?.(e);
+    };
+
+    const handleFocus: React.FocusEventHandler<HTMLInputElement> = (e) => {
+        setActive(true);
+        onFocus?.(e);
+    };
+
+    const handleBlur: React.FocusEventHandler<HTMLInputElement> = (e) => {
+        setActive(false);
+        if (clamp && value) {
+            const clampedValue = clampToNearestStepValue({
+                value,
+                step: baseStep,
+                min,
+                max,
+            });
+            if (value !== clampedValue) {
+                setValue?.(clampedValue);
+                setInputValue(clampedValue.toString());
+            }
+        }
+        onBlur?.(e);
+    };
+
+    const handleUpdate = (v: string) => {
+        setInputValue(v);
+        const preparedStringValue = getPossibleNumberSubstring(v, allowDecimal);
+        updateCursorPosition(innerControlRef, v, preparedStringValue);
+        const {valid, value: parsedNumberValue} = getParsedValue(preparedStringValue);
+        if (valid && parsedNumberValue !== value) {
+            setValue?.(parsedNumberValue);
+        }
+    };
+
+    const handleInput: React.FormEventHandler<HTMLInputElement> = (e) => {
+        const preparedStringValue = getPossibleNumberSubstring(e.currentTarget.value, allowDecimal);
+        updateCursorPosition(innerControlRef, e.currentTarget.value, preparedStringValue);
+    };
+
+    return (
+        <TextInput
+            {...props}
+            className={b({size, view, state}, className)}
+            controlProps={{
+                onInput: handleInput,
+                ...props.controlProps,
+                onWheel: allowMouseWheel && active ? handleWheel : undefined,
+                role: 'spinbutton',
+                inputMode: 'numeric',
```

Comment:
when `allowDecimal = true`, we should pass `decimal` here to have correct keyborad on mobile devices